gcc/rust/typecheck/rust-tyty-rules.h

```diff
@@ -478,6 +486,8 @@ class InferRules : public BaseRules
     BaseRules::visit (type);
   }
 
+  void visit (NeverType &type) override { resolved = base->clone (); }
```

Comment:
I think the only one that should be implementing this should be InferRules

Reply:
But we have to add a new one for NeverType, don't we? After all we cannot use InferType directly because general InferType that cannot be inferred to a concrete type will generate an error message.

And here `unify` is just `unify`, instead of `assign`. Refer to my comment above.

Reply:
InferTypes are inference variables if they uinify against a float/integer they will get a concrete type or if they are unified against a number literal they can then default to i32/f64 respectively.

The only type other than NeverType that should care about NeverType should be general inference variables. At the moment if you have code such as:

```
let a:_ = !
```

a will remain a General Inference variable with your rules here.

If you don't implement the NeverType visitor your rules will correctly give the type mismatch errors which we need at the moment they will be hidden.

Reply:
ignore my comment here i think if we are sure never type works as this then it looks good. Maybe my rust playground examples are not correct since the rust documentation says otherwise.

Reply:
> a will remain a General Inference variable with your rules here.

Maybe surprisingly, it is correct (partially). The below code is tested with rustc.
```rust
fn main() {
    let a: _ = {
        return;
    };
    a = 1;
    a = 1.5; // expected integer, found floating-point number
}
```
But thanks, there is a problem I have not noticed before.
```rust
fn main() {
    let a: _ = {
        return;
    };
}
```
The code above compiles with rustc. But gccrs complains that `a`'s type cannot be determined.

Reply:
Interesting that's going to be a difficult one to fix with the code at present. I have a patch in my PR that should help with that:

https://github.com/Rust-GCC/gccrs/pull/358/commits/3087d72eae82d75b3b4e1abaf35cdeabda5546cc

I think lets fix that test case in a different PR. Ignore my review comments I think you are correct sorry for the questions.

Lets create a separate issue for the never type to track this work: https://github.com/Rust-GCC/gccrs/issues/367

As for this PR i think if you can fix the 3 xfail test cases and add a few new test cases this looks good to me to be merged.

Reply:
I think there must be a case that rustc must default a to be a (). I think we could amend my patch in https://github.com/Rust-GCC/gccrs/commit/3087d72eae82d75b3b4e1abaf35cdeabda5546cc to do that for left over general inference variables. At the end of type resolution in the name scope.

Reply:
> As for this PR i think if you can fix the 3 xfail test cases and add a few new test cases this looks good to me to be merged.

Thanks. Maybe I'll do this tomorrow. Tonight (UTC+8) let me get my GSoC proposal done first. It is really close to the deadline (sorry for the late).

Reply:
> I think there must be a case that rustc must default a to be a (). I think we could amend my patch in 3087d72 to do that for left over general inference variables. At the end of type resolution in the name scope.

And this does NOT compile.
```rust
fn main() {
    let a; // consider giving `a` a type
}
```